src/Workspaces/Core/Portable/Options/OptionServiceFactory.cs

```diff
@@ -0,0 +1,121 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+
+namespace Microsoft.CodeAnalysis.Options
+{
+    [ExportWorkspaceServiceFactory(typeof(IOptionService)), Shared]
+    internal class OptionServiceFactory : IWorkspaceServiceFactory
+    {
+        private readonly IGlobalOptionService _globalOptionService;
+
+        [ImportingConstructor]
+        public OptionServiceFactory(IGlobalOptionService globalOptionService)
+        {
+            _globalOptionService = globalOptionService;
+        }
+
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+        {
+            return new OptionService(_globalOptionService, workspaceServices);
+        }
+
+        /// <summary>
+        /// Wraps an underlying <see cref="IGlobalOptionService"/> and exposes its data to workspace
+        /// clients.  Also takes the <see cref="IGlobalOptionService.OptionChanged"/> notifications
+        /// and forwards them along using the same <see cref="IWorkspaceTaskScheduler"/> used by the
+        /// <see cref="Workspace"/> this is connected to.  i.e. instead of synchronously just passing
+        /// along the underlying events, these will be enqueued onto the workspaces eventing queue.
+        /// </summary>
+        // Internal for testing purposes.
+        internal class OptionService : IWorkspaceOptionService
+        {
+            private readonly IGlobalOptionService _globalOptionService;
+
+            // Can be null during testing.
+            private readonly IWorkspaceTaskScheduler _taskQueue;
+
+            private readonly object _gate = new object();
+            private ImmutableArray<EventHandler<OptionChangedEventArgs>> _eventHandlers =
+                ImmutableArray<EventHandler<OptionChangedEventArgs>>.Empty;
+
+            public OptionService(
+                IGlobalOptionService globalOptionService,
+                HostWorkspaceServices workspaceServices)
+            {
+                _globalOptionService = globalOptionService;
+
+                var workspaceTaskSchedulerFactory = workspaceServices?.GetRequiredService<IWorkspaceTaskSchedulerFactory>();
+                _taskQueue = workspaceTaskSchedulerFactory?.CreateTaskQueue();
+
+                _globalOptionService.OptionChanged += OnGlobalOptionServiceOptionChanged;
+            }
+
+            public void OnWorkspaceDisposed(Workspace workspace)
```

Comment:
As a general pattern, should we just allow IWorkspaceServices to optionally implement IDisposable? I'm reminded of the pattern for ITaggers and IClassifiers in the editor. I'm guessing we have other one-off services that might want this as well.